File: flutter/flutter-sdk-bridge/ping_oidc/ios/ping_oidc/Sources/ping_oidc/OidcClientFactory.swift
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import Foundation
import PingOidc
import ping_core

/// Wraps a live native `OidcClient` plus the flat `OidcConfigMessage` it was built from, so it
/// can be stored in `CoreRuntime.oidcClientRegistry` and read back through `OidcConfigHandle` by
/// consumers with no compile-time dependency on `ping_oidc`. `@unchecked Sendable` because
/// `OidcClient` is not itself `Sendable` — mirrors `ping_journey`'s `JourneyHandle`.
final class OidcClientHandle: NativeHandle, OidcConfigHandle, @unchecked Sendable {
    let payload: OidcConfigMessage
    let client: OidcClient

    init(payload: OidcConfigMessage, client: OidcClient) {
        self.payload = payload
        self.client = client
    }

    var clientId: String { payload.clientId }
    var discoveryEndpoint: String? { payload.discoveryEndpoint }
    var openId: OidcOpenIdConfig? {
        guard let openId = payload.openId else { return nil }
        return OidcOpenIdConfig(
            authorizationEndpoint: openId.authorizationEndpoint,
            tokenEndpoint: openId.tokenEndpoint,
            userinfoEndpoint: openId.userinfoEndpoint,
            endSessionEndpoint: openId.endSessionEndpoint,
            pingEndIdpSessionEndpoint: openId.pingEndIdpSessionEndpoint,
            revocationEndpoint: openId.revocationEndpoint
        )
    }
    var redirectUri: String? { payload.redirectUri }
    var scopes: [String] { payload.scopes?.compactMap { $0 } ?? [] }
    var acrValues: String? { payload.acrValues }
    var signOutRedirectUri: String? { nil }  // no such property on iOS's OidcClientConfig
    var state: String? { payload.state }
    var nonce: String? { payload.nonce }
    var uiLocales: String? { payload.uiLocales }
    var refreshThreshold: Int64? { payload.refreshThresholdSeconds }
    var loginHint: String? { payload.loginHint }
    var display: String? { payload.display }
    var prompt: String? { payload.prompt }
    var additionalParameters: [String: String] {
        payload.additionalParameters?.reduce(into: [String: String]()) { result, entry in
            if let key = entry.key, let value = entry.value { result[key] = value }
        } ?? [:]
    }
    var par: Bool { payload.par }
}

/// Builds a native `OidcClient` from an `OidcConfigMessage` and registers it in the shared core
/// registry, returning the generated handle id.
enum OidcClientFactory {
    static func create(_ config: OidcConfigMessage) async throws -> String {
        let nativeConfig = try OidcConfigParser.parse(config)
        let client = OidcClient(config: nativeConfig)
        let handle = OidcClientHandle(payload: config, client: client)
        return await CoreRuntime.oidcClientRegistry.register(handle)
    }
}

File: flutter/flutter-sdk-bridge/ping_oidc/android/src/main/kotlin/com/pingidentity/flutter/oidc/OidcClientFactory.kt
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package com.pingidentity.flutter.oidc

import com.pingidentity.flutter.core.CoreRuntime
import com.pingidentity.flutter.core.oidc.OidcConfigHandle
import com.pingidentity.flutter.core.oidc.OidcOpenIdConfig
import com.pingidentity.flutter.core.registry.NativeHandle
import com.pingidentity.oidc.OidcClient

/**
 * Wraps a live native [OidcClient] plus the flat [OidcConfigMessage] it was built from, so it can
 * be stored in [CoreRuntime.oidcClientRegistry] and read back through [OidcConfigHandle] by
 * consumers with no compile-time dependency on `ping_oidc` (e.g. `ping_journey`'s Phase 6
 * delegation).
 */
internal class OidcClientHandle(
    val payload: OidcConfigMessage,
    val client: OidcClient,
) : NativeHandle, OidcConfigHandle {
    override val clientId: String get() = payload.clientId
    override val discoveryEndpoint: String? get() = payload.discoveryEndpoint
    override val openId: OidcOpenIdConfig?
        get() =
            payload.openId?.let {
                OidcOpenIdConfig(
                    authorizationEndpoint = it.authorizationEndpoint,
                    tokenEndpoint = it.tokenEndpoint,
                    userinfoEndpoint = it.userinfoEndpoint,
                    endSessionEndpoint = it.endSessionEndpoint,
                    pingEndIdpSessionEndpoint = it.pingEndIdpSessionEndpoint,
                    revocationEndpoint = it.revocationEndpoint,
                )
            }
    override val redirectUri: String? get() = payload.redirectUri
    override val scopes: List<String> get() = payload.scopes?.filterNotNull() ?: emptyList()
    override val acrValues: String? get() = payload.acrValues
    override val signOutRedirectUri: String? get() = payload.signOutRedirectUri
    override val state: String? get() = payload.state
    override val nonce: String? get() = payload.nonce
    override val uiLocales: String? get() = payload.uiLocales
    override val refreshThreshold: Long? get() = payload.refreshThresholdSeconds
    override val loginHint: String? get() = payload.loginHint
    override val display: String? get() = payload.display
    override val prompt: String? get() = payload.prompt
    override val additionalParameters: Map<String, String>
        get() =
            payload.additionalParameters
                ?.entries
                ?.filter { it.key != null && it.value != null }
                ?.associate { it.key!! to it.value!! } ?: emptyMap()
    override val par: Boolean get() = payload.par
}

/**
 * Builds a native [OidcClient] from an [OidcConfigMessage] and registers it in the shared core
 * registry, returning the generated handle id.
 */
internal object OidcClientFactory {
    fun create(config: OidcConfigMessage): String {
        val nativeConfig = OidcConfigParser.parse(config)
        val client = OidcClient(nativeConfig)
        return CoreRuntime.oidcClientRegistry.register(OidcClientHandle(config, client))
    }
}
